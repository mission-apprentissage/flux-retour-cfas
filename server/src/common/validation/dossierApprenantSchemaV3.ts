import { z } from "zod";

import { primitivesV1, primitivesV3 } from "@/common/validation/utils/zodPrimitives";

export const dossierApprenantSchemaV3Base = () =>
  z.object({
    // REQUIRED FIELDS
    nom_apprenant: primitivesV1.apprenant.nom,
    prenom_apprenant: primitivesV1.apprenant.prenom,
    date_de_naissance_apprenant: primitivesV1.apprenant.date_de_naissance,
    // The following fields are missing in V3
    // uai_etablissement: primitivesV1.etablissement_responsable.uai,
    // nom_etablissement: primitivesV1.etablissement_responsable.nom,
    // id_formation: primitivesV1.formation.code_cfd,
    annee_scolaire: primitivesV1.formation.annee_scolaire,
    statut_apprenant: primitivesV1.apprenant.statut,
    date_metier_mise_a_jour_statut: primitivesV1.apprenant.date_metier_mise_a_jour_statut,
    id_erp_apprenant: primitivesV1.apprenant.id_erp,

    // OPTIONAL FIELDS
    ine_apprenant: primitivesV1.apprenant.ine.optional(),
    email_contact: primitivesV1.apprenant.email.optional(),
    tel_apprenant: primitivesV1.apprenant.telephone.nullish(),
    // The following field is missing in V3
    // siret_etablissement: primitivesV1.etablissement_responsable.siret.optional(),
    // code_commune_insee_apprenant: primitivesV1.apprenant.code_commune_insee.optional(),
    libelle_court_formation: primitivesV1.formation.libelle_court.optional(),
    // The following field are missing in V3
    // libelle_long_formation: primitivesV1.formation.libelle_long.optional(),
    // periode_formation: primitivesV1.formation.periode.nullish(),
    annee_formation: primitivesV1.formation.annee.optional(),
    formation_rncp: primitivesV1.formation.code_rncp.optional(),
    contrat_date_debut: primitivesV1.contrat.date_debut.optional(),
    contrat_date_fin: primitivesV1.contrat.date_fin.optional(),
    contrat_date_rupture: primitivesV1.contrat.date_rupture.optional(),

    // V3 FIELDS
    // OPTIONAL FIELDS
    nir_apprenant: primitivesV3.apprenant.nir.optional(),
    adresse_apprenant: primitivesV3.apprenant.adresse.optional(),
    code_postal_apprenant: primitivesV3.apprenant.code_postal.optional(),
    code_postal_de_naissance_apprenant: primitivesV3.apprenant.code_postal.optional(),
    sexe_apprenant: primitivesV3.apprenant.sexe.optional(),
    rqth_apprenant: primitivesV3.apprenant.rqth.optional(),
    date_rqth_apprenant: primitivesV3.apprenant.date_rqth.optional(),
    responsable_apprenant_mail1: primitivesV3.responsable.email.optional(),
    responsable_apprenant_mail2: primitivesV3.responsable.email.optional(),
    obtention_diplome_formation: primitivesV3.formation.obtention_diplome.optional(),
    date_obtention_diplome_formation: primitivesV3.formation.date_obtention_diplome.optional(),
    date_exclusion_formation: primitivesV3.formation.date_exclusion.optional(),
    cause_exclusion_formation: primitivesV3.formation.cause_exclusion.optional(),
    nom_referent_handicap_formation: primitivesV3.formation.referent_handicap.nom.optional(),
    prenom_referent_handicap_formation: primitivesV3.formation.referent_handicap.prenom.optional(),
    email_referent_handicap_formation: primitivesV3.formation.referent_handicap.email.optional(),
    cause_rupture_contrat: primitivesV3.contrat.cause_rupture.optional(),
    contrat_date_debut_2: primitivesV1.contrat.date_debut.optional(),
    contrat_date_fin_2: primitivesV1.contrat.date_fin.optional(),
    contrat_date_rupture_2: primitivesV1.contrat.date_rupture.optional(),
    cause_rupture_contrat_2: primitivesV3.contrat.cause_rupture.optional(),
    contrat_date_debut_3: primitivesV1.contrat.date_debut.optional(),
    contrat_date_fin_3: primitivesV1.contrat.date_fin.optional(),
    contrat_date_rupture_3: primitivesV1.contrat.date_rupture.optional(),
    cause_rupture_contrat_3: primitivesV3.contrat.cause_rupture.optional(),
    contrat_date_debut_4: primitivesV1.contrat.date_debut.optional(),
    contrat_date_fin_4: primitivesV1.contrat.date_fin.optional(),
    contrat_date_rupture_4: primitivesV1.contrat.date_rupture.optional(),
    cause_rupture_contrat_4: primitivesV3.contrat.cause_rupture.optional(),
    siret_employeur: primitivesV3.employeur.siret.optional(),
    siret_employeur_2: primitivesV3.employeur.siret.optional(),
    siret_employeur_3: primitivesV3.employeur.siret.optional(),
    siret_employeur_4: primitivesV3.employeur.siret.optional(),
    formation_presentielle: primitivesV3.formation.presentielle.optional(),
    // These two fields should have been required but sometimes, it is missing in YMAG
    duree_theorique_formation: primitivesV3.formation.duree_theorique.optional(), // Legacy, but still in use by ERPs and Excel import.
    duree_theorique_formation_mois: primitivesV3.formation.duree_theorique_mois.optional(), // The new field.

    // REQUIRED FIELDS
    date_inscription_formation: primitivesV3.formation.date_inscription,
    date_entree_formation: primitivesV3.formation.date_entree,
    date_fin_formation: primitivesV3.formation.date_fin,

    etablissement_responsable_uai: primitivesV1.etablissement_responsable.uai,
    etablissement_responsable_siret: primitivesV1.etablissement_responsable.siret,
    etablissement_formateur_uai: primitivesV1.etablissement_formateur.uai,
    etablissement_formateur_siret: primitivesV1.etablissement_formateur.siret,
    etablissement_lieu_de_formation_uai: primitivesV1.etablissement_lieu_de_formation.uai,
    etablissement_lieu_de_formation_siret: primitivesV1.etablissement_lieu_de_formation.siret,

    formation_cfd: primitivesV1.formation.code_cfd.optional(),
    // Champs SIFA
    derniere_situation: primitivesV3.derniere_situation.optional(),
    dernier_organisme_uai: primitivesV3.dernier_organisme_uai.optional(),
    type_cfa: primitivesV3.type_cfa.optional(),
  });

const dossierApprenantSchemaV3 = () => {
  return dossierApprenantSchemaV3Base().merge(
    z.object({
      // These fields are hidden in documentation because they are generated by the API itself.
      api_version: primitivesV1.api_version.optional(),
      source: primitivesV1.source,
      source_organisme_id: primitivesV1.source_organisme_id.optional(),
    })
  );
};

// Utilisé par le téléversement manuel. Le but côté métier est d'obliger les utilisateurs à remplir des champs dont
// ils pourraient avoir besoin dans SIFA alors que ce n'est pas toujours présent côté ERPs.
// Source de la décision: https://mission-apprentissage.slack.com/archives/C02FR2L1VB8/p1693232432865229
export const dossierApprenantSchemaV3WithMoreRequiredFields = () => {
  return dossierApprenantSchemaV3Base().merge(
    z.object({
      email_contact: primitivesV1.apprenant.email,
      adresse_apprenant: primitivesV3.apprenant.adresse,
      code_postal_apprenant: primitivesV3.apprenant.code_postal,
      sexe_apprenant: primitivesV3.apprenant.sexe,
      annee_formation: primitivesV1.formation.annee,
    })
  );
};

export function dossierApprenantSchemaV3WithMoreRequiredFieldsValidatingUAISiret(
  invalidsUais: string[],
  invalidsSirets: string[]
) {
  const validateUAI = (uai: string) => !invalidsUais.includes(uai);
  const validateSiret = (siret: string) => !invalidsSirets.includes(siret);
  const messageUai = "UAI non valide";
  const messageSiret = "Siret non valide";

  return dossierApprenantSchemaV3WithMoreRequiredFields().merge(
    z.object({
      etablissement_responsable_uai: primitivesV1.etablissement_responsable.uai.refine(validateUAI, {
        message: messageUai,
      }),
      etablissement_responsable_siret: primitivesV1.etablissement_responsable.siret.refine(validateSiret, {
        message: messageSiret,
      }),
      etablissement_formateur_uai: primitivesV1.etablissement_formateur.uai.refine(validateUAI, {
        message: messageUai,
      }),
      etablissement_formateur_siret: primitivesV1.etablissement_formateur.siret.refine(validateSiret, {
        message: messageSiret,
      }),
      etablissement_lieu_de_formation_uai: primitivesV1.etablissement_lieu_de_formation.uai.refine(validateUAI, {
        message: messageUai,
      }),
      etablissement_lieu_de_formation_siret: primitivesV1.etablissement_lieu_de_formation.siret.refine(validateSiret, {
        message: messageSiret,
      }),
    })
  );
}

export type DossierApprenantSchemaV3ZodType = z.input<ReturnType<typeof dossierApprenantSchemaV3>>;

export default dossierApprenantSchemaV3;
