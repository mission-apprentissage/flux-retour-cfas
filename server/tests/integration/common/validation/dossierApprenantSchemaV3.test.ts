import dossierApprenantSchemaV3, {
  DossierApprenantSchemaV3ZodType,
} from "@/common/validation/dossierApprenantSchemaV3";

const validInput: DossierApprenantSchemaV3ZodType = {
  source: "test",
  apprenant: {
    nom: "test",
    prenom: "test",
    date_de_naissance: new Date().toISOString(),
    statut: 1,
    date_metier_mise_a_jour_statut: new Date().toISOString(),
    id_erp: "test",
  },
  etablissement_responsable: {
    siret: "12345678912345",
  },
  etablissement_formateur: {
    siret: "12345678912345",
  },
  formation: {
    code_cfd: "50022141",
    code_rncp: "RNCP35316",
    periode: "2022-2023",
    annee_scolaire: "2022-2023",
    date_inscription: new Date().toISOString(),
    date_entree: new Date().toISOString(),
    date_fin: new Date().toISOString(),
  },
};

const expected = {
  source: "test",
  apprenant: {
    nom: "TEST",
    prenom: "test",
    date_de_naissance: expect.any(Date),
    statut: 1,
    date_metier_mise_a_jour_statut: expect.any(Date),
    id_erp: "test",
  },
  etablissement_responsable: {
    siret: "12345678912345",
  },
  etablissement_formateur: {
    siret: "12345678912345",
  },
  formation: {
    code_cfd: "50022141",
    code_rncp: "RNCP35316",
    periode: [2022, 2023],
    annee_scolaire: "2022-2023",
    date_inscription: expect.any(Date),
    date_entree: expect.any(Date),
    date_fin: expect.any(Date),
  },
};

const invalidInput = {
  source: null,
  apprenant: {
    nom: 1,
    prenom: 2,
    date_de_naissance: 3,
    statut: "?",
    date_metier_mise_a_jour_statut: "2002/02/02",
    id_erp: true,
    // V1 - OPTIONAL FIELDS
    ine: true,
    email: "test",
    telephone: "abc",
    code_commune_insee: "abc",
    // V3 - OPTIONAL FIELDS
    sexe: "homme",
    rqth: "non",
    date_rqth: "02/02/2002",
  },
  etablissement_responsable: {
    siret: "1",
    uai: "2",
    nom: "",
  },
  etablissement_formateur: {
    siret: "test",
    uai: "test",
    nom: true,
    code_commune_insee: "test",
  },
  formation: {
    libelle_court: "",
    libelle_long: "",
    periode: "1000-20000",
    annee_scolaire: "2100-2200",
    annee: 100,
    code_rncp: "test",
    code_cfd: "test",
    // V3 - REQUIRED FIELDS
    date_inscription: "20/02/07",
    date_entree: "07-02-20",
    date_fin: "2022-02",
    // V3 - OPTIONAL FIELDS
    obtention_diplome: "test",
    date_obtention_diplome: "test",
    date_exclusion: "test",
    cause_exclusion: 1,
    referent_handicap: {
      nom: 1,
      prenom: 2,
      email: "test",
    },
  },
  contrat: {
    date_debut: "test",
    date_fin: 2000,
    date_rupture: "test",
    // V3 - OPTIONAL FIELDS
    cause_rupture: 1,
  },
  employeur: {
    siret: true,
    code_commune_insee: true,
    code_naf: "test",
  },
};

describe("dossierApprenantSchemaV3.test", () => {
  it("returns required root fields with empty entry", () => {
    const output = dossierApprenantSchemaV3().strict().safeParse({});
    if (output.success) {
      throw new Error("should not be a success");
    }
    expect(output.error.issues.map((o) => o)).toMatchSnapshot();
  });

  it("returns required fields with empty entry", () => {
    const output = dossierApprenantSchemaV3()
      .strict()
      .safeParse({ apprenant: {}, etablissement_responsable: {}, etablissement_formateur: {}, formation: {} });
    if (output.success) {
      throw new Error("should not be a success");
    }
    expect(output.error.issues.map((o) => o)).toMatchSnapshot();
  });

  it("return all errors with a complete invalid entry", () => {
    const output = dossierApprenantSchemaV3().strict().safeParse(invalidInput);

    if (output.success) {
      throw new Error("should not be a success");
    }
    expect(output.error.issues.map((o) => o)).toMatchSnapshot();
  });

  it("accept a valid dossierApprenantV3", () => {
    const output = dossierApprenantSchemaV3().strict().safeParse(validInput);

    expect(output).toStrictEqual({ success: true, data: expected });
  });

  it("is idempotent", () => {
    const output = dossierApprenantSchemaV3().strict().safeParse(validInput);
    if (!output.success) {
      throw new Error("output is not success");
    }
    const output2 = dossierApprenantSchemaV3().strict().safeParse(output.data);
    expect(output2).toStrictEqual({ success: true, data: expected });
  });
});
