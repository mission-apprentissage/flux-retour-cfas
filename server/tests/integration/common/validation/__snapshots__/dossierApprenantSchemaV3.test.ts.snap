// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dossierApprenantSchemaV3.test return all errors with a complete invalid entry 1`] = `
[
  {
    "code": "invalid_type",
    "expected": "string",
    "message": "String attendu",
    "path": [
      "source",
    ],
    "received": "null",
  },
  {
    "code": "invalid_type",
    "expected": "string",
    "message": "String attendu",
    "path": [
      "apprenant",
      "nom",
    ],
    "received": "number",
  },
  {
    "code": "invalid_type",
    "expected": "string",
    "message": "String attendu",
    "path": [
      "apprenant",
      "prenom",
    ],
    "received": "number",
  },
  {
    "code": "invalid_type",
    "expected": "date",
    "message": "Date invalide",
    "path": [
      "apprenant",
      "date_de_naissance",
    ],
    "received": "number",
  },
  {
    "code": "invalid_type",
    "expected": "number",
    "message": "Valeurs possibles: 0,2,3",
    "path": [
      "apprenant",
      "statut",
    ],
    "received": "string",
  },
  {
    "code": "invalid_type",
    "expected": "date",
    "message": "Date invalide",
    "path": [
      "apprenant",
      "date_metier_mise_a_jour_statut",
    ],
    "received": "string",
  },
  {
    "code": "invalid_type",
    "expected": "string",
    "message": "String attendu",
    "path": [
      "apprenant",
      "id_erp",
    ],
    "received": "boolean",
  },
  {
    "code": "invalid_type",
    "expected": "string",
    "message": "String attendu",
    "path": [
      "apprenant",
      "ine",
    ],
    "received": "boolean",
  },
  {
    "code": "invalid_string",
    "message": "Email non valide",
    "path": [
      "apprenant",
      "email",
    ],
    "validation": "email",
  },
  {
    "code": "invalid_string",
    "message": "Format invalide",
    "path": [
      "apprenant",
      "telephone",
    ],
    "validation": "regex",
  },
  {
    "code": "invalid_string",
    "message": "Format invalide",
    "path": [
      "apprenant",
      "code_commune_insee",
    ],
    "validation": "regex",
  },
  {
    "code": "invalid_type",
    "expected": "boolean",
    "message": "Boolean attendu",
    "path": [
      "apprenant",
      "rqth",
    ],
    "received": "string",
  },
  {
    "code": "invalid_type",
    "expected": "date",
    "message": "Date invalide",
    "path": [
      "apprenant",
      "date_rqth",
    ],
    "received": "string",
  },
  {
    "code": "invalid_string",
    "message": "SIRET invalide",
    "path": [
      "etablissement_responsable",
      "siret",
    ],
    "validation": "regex",
  },
  {
    "code": "invalid_string",
    "message": "UAI invalide",
    "path": [
      "etablissement_responsable",
      "uai",
    ],
    "validation": "regex",
  },
  {
    "code": "invalid_string",
    "message": "SIRET invalide",
    "path": [
      "etablissement_formateur",
      "siret",
    ],
    "validation": "regex",
  },
  {
    "code": "invalid_string",
    "message": "UAI invalide",
    "path": [
      "etablissement_formateur",
      "uai",
    ],
    "validation": "regex",
  },
  {
    "code": "invalid_type",
    "expected": "string",
    "message": "String attendu",
    "path": [
      "etablissement_formateur",
      "nom",
    ],
    "received": "boolean",
  },
  {
    "code": "invalid_string",
    "message": "Format invalide",
    "path": [
      "etablissement_formateur",
      "code_commune_insee",
    ],
    "validation": "regex",
  },
  {
    "code": "too_small",
    "exact": false,
    "inclusive": true,
    "message": "String must contain at least 2 character(s)",
    "minimum": 2,
    "path": [
      "formation",
      "libelle_court",
    ],
    "type": "string",
  },
  {
    "code": "too_small",
    "exact": false,
    "inclusive": true,
    "message": "String must contain at least 2 character(s)",
    "minimum": 2,
    "path": [
      "formation",
      "libelle_long",
    ],
    "type": "string",
  },
  {
    "code": "too_small",
    "exact": false,
    "inclusive": true,
    "message": "Number must be greater than or equal to 2000",
    "minimum": 2000,
    "path": [
      "formation",
      "periode",
      0,
    ],
    "type": "number",
  },
  {
    "code": "too_big",
    "exact": false,
    "inclusive": true,
    "maximum": 2100,
    "message": "Number must be less than or equal to 2100",
    "path": [
      "formation",
      "periode",
      1,
    ],
    "type": "number",
  },
  {
    "code": "too_big",
    "exact": false,
    "inclusive": true,
    "maximum": 5,
    "message": "Number must be less than or equal to 5",
    "path": [
      "formation",
      "annee",
    ],
    "type": "number",
  },
  {
    "code": "invalid_string",
    "message": "Code RNCP invalide",
    "path": [
      "formation",
      "code_rncp",
    ],
    "validation": "regex",
  },
  {
    "code": "invalid_string",
    "message": "Code CFD invalide",
    "path": [
      "formation",
      "code_cfd",
    ],
    "validation": "regex",
  },
  {
    "code": "invalid_type",
    "expected": "date",
    "message": "Date invalide",
    "path": [
      "formation",
      "date_inscription",
    ],
    "received": "string",
  },
  {
    "code": "invalid_type",
    "expected": "date",
    "message": "Date invalide",
    "path": [
      "formation",
      "date_entree",
    ],
    "received": "string",
  },
  {
    "code": "invalid_type",
    "expected": "date",
    "message": "Date invalide",
    "path": [
      "formation",
      "date_fin",
    ],
    "received": "string",
  },
  {
    "code": "invalid_type",
    "expected": "boolean",
    "message": "Boolean attendu",
    "path": [
      "formation",
      "obtention_diplome",
    ],
    "received": "string",
  },
  {
    "code": "invalid_type",
    "expected": "date",
    "message": "Date invalide",
    "path": [
      "formation",
      "date_obtention_diplome",
    ],
    "received": "string",
  },
  {
    "code": "invalid_type",
    "expected": "date",
    "message": "Date invalide",
    "path": [
      "formation",
      "date_exclusion",
    ],
    "received": "string",
  },
  {
    "code": "invalid_type",
    "expected": "string",
    "message": "String attendu",
    "path": [
      "formation",
      "cause_exclusion",
    ],
    "received": "number",
  },
  {
    "code": "invalid_type",
    "expected": "string",
    "message": "String attendu",
    "path": [
      "formation",
      "referent_handicap",
      "nom",
    ],
    "received": "number",
  },
  {
    "code": "invalid_type",
    "expected": "string",
    "message": "String attendu",
    "path": [
      "formation",
      "referent_handicap",
      "prenom",
    ],
    "received": "number",
  },
  {
    "code": "invalid_string",
    "message": "Invalid email",
    "path": [
      "formation",
      "referent_handicap",
      "email",
    ],
    "validation": "email",
  },
  {
    "code": "invalid_type",
    "expected": "date",
    "message": "Date invalide",
    "path": [
      "contrat",
      "date_debut",
    ],
    "received": "string",
  },
  {
    "code": "invalid_type",
    "expected": "date",
    "message": "Date invalide",
    "path": [
      "contrat",
      "date_fin",
    ],
    "received": "number",
  },
  {
    "code": "invalid_type",
    "expected": "date",
    "message": "Date invalide",
    "path": [
      "contrat",
      "date_rupture",
    ],
    "received": "string",
  },
  {
    "code": "invalid_type",
    "expected": "string",
    "message": "String attendu",
    "path": [
      "contrat",
      "cause_rupture",
    ],
    "received": "number",
  },
  {
    "code": "invalid_type",
    "expected": "string",
    "message": "String attendu",
    "path": [
      "employeur",
      "siret",
    ],
    "received": "boolean",
  },
  {
    "code": "invalid_type",
    "expected": "string",
    "message": "String attendu",
    "path": [
      "employeur",
      "code_commune_insee",
    ],
    "received": "boolean",
  },
  {
    "code": "invalid_string",
    "message": "UAI invalide",
    "path": [
      "employeur",
      "code_naf",
    ],
    "validation": "regex",
  },
]
`;

exports[`dossierApprenantSchemaV3.test returns required fields with empty entry 1`] = `
[
  {
    "code": "invalid_type",
    "expected": "string",
    "message": "String attendu",
    "path": [
      "source",
    ],
    "received": "undefined",
  },
  {
    "code": "invalid_type",
    "expected": "string",
    "message": "String attendu",
    "path": [
      "apprenant",
      "nom",
    ],
    "received": "undefined",
  },
  {
    "code": "invalid_type",
    "expected": "string",
    "message": "String attendu",
    "path": [
      "apprenant",
      "prenom",
    ],
    "received": "undefined",
  },
  {
    "code": "invalid_type",
    "expected": "date",
    "message": "Champ obligatoire",
    "path": [
      "apprenant",
      "date_de_naissance",
    ],
    "received": "undefined",
  },
  {
    "code": "invalid_type",
    "expected": "number",
    "message": "Number attendu",
    "path": [
      "apprenant",
      "statut",
    ],
    "received": "undefined",
  },
  {
    "code": "invalid_type",
    "expected": "date",
    "message": "Champ obligatoire",
    "path": [
      "apprenant",
      "date_metier_mise_a_jour_statut",
    ],
    "received": "undefined",
  },
  {
    "code": "invalid_type",
    "expected": "string",
    "message": "String attendu",
    "path": [
      "apprenant",
      "id_erp",
    ],
    "received": "undefined",
  },
  {
    "code": "custom",
    "message": "SIRET ou UAI requis",
    "path": [
      "etablissement_responsable",
    ],
  },
  {
    "code": "custom",
    "message": "SIRET ou UAI requis",
    "path": [
      "etablissement_formateur",
    ],
  },
  {
    "code": "invalid_type",
    "expected": "string",
    "message": "String attendu",
    "path": [
      "formation",
      "annee_scolaire",
    ],
    "received": "undefined",
  },
  {
    "code": "invalid_type",
    "expected": "date",
    "message": "Champ obligatoire",
    "path": [
      "formation",
      "date_inscription",
    ],
    "received": "undefined",
  },
  {
    "code": "invalid_type",
    "expected": "date",
    "message": "Champ obligatoire",
    "path": [
      "formation",
      "date_entree",
    ],
    "received": "undefined",
  },
  {
    "code": "invalid_type",
    "expected": "date",
    "message": "Champ obligatoire",
    "path": [
      "formation",
      "date_fin",
    ],
    "received": "undefined",
  },
]
`;

exports[`dossierApprenantSchemaV3.test returns required root fields with empty entry 1`] = `
[
  {
    "code": "invalid_type",
    "expected": "string",
    "message": "String attendu",
    "path": [
      "source",
    ],
    "received": "undefined",
  },
  {
    "code": "invalid_type",
    "expected": "object",
    "message": "Object attendu",
    "path": [
      "apprenant",
    ],
    "received": "undefined",
  },
  {
    "code": "invalid_type",
    "expected": "object",
    "message": "Object attendu",
    "path": [
      "etablissement_responsable",
    ],
    "received": "undefined",
  },
  {
    "code": "invalid_type",
    "expected": "object",
    "message": "Object attendu",
    "path": [
      "etablissement_formateur",
    ],
    "received": "undefined",
  },
  {
    "code": "invalid_type",
    "expected": "object",
    "message": "Object attendu",
    "path": [
      "formation",
    ],
    "received": "undefined",
  },
]
`;
